pkg/datapath/loader/compile.go

```diff
@@ -0,0 +1,182 @@
+
+package loader
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"fmt"
+	"os"
+	"os/exec"
+	"runtime"
+)
+
+// OutputType determines the type to be generated by the compilation steps.
+type OutputType string
+
+const (
+	outputObject   = OutputType("obj")
+	outputAssembly = OutputType("asm")
+	outputSource   = OutputType("c")
+)
+
+// progInfo describes a program to be compiled with the expected output format
+type progInfo struct {
+	// Source is the program source (base) filename to be compiled
+	Source string
+	// Output is the expected (base) filename produced from the source
+	Output string
+	// OutputType to be created by LLVM
+	OutputType OutputType
+}
+
+// directoryInfo includes relevant directories for compilation and linking
+type directoryInfo struct {
+	// Library contains the library code to be used for compilation
+	Library string
+	// Runtime contains headers for compilation
+	Runtime string
+	// Output is the directory where the files will be stored
+	Output string
+}
+
+var (
+	compiler       = "clang"
+	linker         = "llc"
+	standardCFlags = []string{"-O2", "-target", "bpf",
+		fmt.Sprintf("-D__NR_CPUS__=%d", runtime.NumCPU()),
+		"-Wno-address-of-packed-member", "-Wno-unknown-warning-option"}
+	standardLDFlags = []string{"-march=bpf", "-mcpu=probe"}
+
+	endpointPrefix   = "bpf_lxc"
+	endpointProg     = fmt.Sprintf("%s.%s", endpointPrefix, outputSource)
+	endpointObj      = fmt.Sprintf("%s.%s", endpointPrefix, outputObject)
+	endpointObjDebug = fmt.Sprintf("%s.dbg.%s", endpointPrefix, outputObject)
+	endpointAsm      = fmt.Sprintf("%s.%s", endpointPrefix, outputAssembly)
+
+	debugProgs = []*progInfo{
+		{
+			Source:     endpointProg,
+			Output:     endpointObjDebug,
+			OutputType: outputObject,
+		},
+		{
+			Source:     endpointProg,
+			Output:     endpointAsm,
+			OutputType: outputAssembly,
+		},
+		{
+			Source:     endpointProg,
+			Output:     endpointProg,
+			OutputType: outputSource,
+		},
+	}
+	datapathProg = &progInfo{
+		Source:     endpointProg,
+		Output:     endpointObj,
+		OutputType: outputObject,
+	}
+)
+
+// irPath determines a temporary path on the filesystem to store the compiled
+// intermediate representation object for the program.
+//
+// TODO: Remove this and just pipe the 'clang' and 'llc' programs together.
+func irPath(prog *progInfo, dir *directoryInfo) string {
+	return fmt.Sprintf("%s/%s.bc", dir.Output, prog.Output)
+}
+
+// progLDFlags determines the loader flags for the specified prog and paths.
+func progLDFlags(prog *progInfo, dir *directoryInfo) []string {
+	return []string{
+		fmt.Sprintf("-filetype=%s", prog.OutputType),
+		irPath(prog, dir),
+		"-o", fmt.Sprintf("%s/%s", dir.Output, prog.Output),
+	}
+}
+
+// Link the specified program from the specified path to the intermediate
+// representation, to the output specified in the prog's info.
+func Link(ctx context.Context, prog *progInfo, dir *directoryInfo, debug bool) error {
+	args := make([]string, 0, 8)
```

Comment:
Where does the 8 comes from?

Reply:
We typically use less than 8 ld flags. I looked at how many we used, and picked a number. It doesn't really matter, and the capacity will expand if the number isn't right, but we also don't change these very often.

Reply:
@joestringer can you add a comment indicating as such in the code?